DOC Document new non-blocking sessions solution (#733)
silverstripe/developer-docs: en/08_Changelogs/6.1.0.md

---
title: 6.0.0 (unreleased)
---

# 6.0.0 (unreleased)

## Overview

- [Features and enhancements](#features-and-enhancements)
  - [Non-blocking file-based sessions](#non-blocking-sessions)
  - [Other new features](#other-new-features)
- [API changes](#api-changes)
- [Bug fixes](#bug-fixes)

## Features and enhancements

## Non-blocking file-based sessions {#non-blocking-sessions}

> The default file-based session handler for PHP holds a lock on the session file while the session is open. In practice this means that multiple concurrent requests from the same user have to wait for one another to finish processing after a session has been started. This includes AJAX requests.

Reviewer: ```suggestion
The default file-based session handler for PHP holds a lock on the session file while the session is open. This means that multiple concurrent requests from the same user have to wait for one another to finish processing after a session has been started. This includes XHR requests.
```

Reviewer 2: XHR vs AJAX discussion can be held in https://github.com/silverstripe/developer-docs/pull/733#discussion_r2074381494

Reviewer 2: Made the other change.

> To resolve this problem, a new [`Session.save_handler`](api:SilverStripe\Control\Session->save_handler) configuration property has been added. By default this is set to [`FileSessionHandler`](api:SilverStripe\Control\SessionHandler\FileSessionHandler).

Reviewer: ```suggestion
To resolve this problem, a new [`Session.save_handler`](api:SilverStripe\Control\Session->save_handler) configuration property has been added. By default this is set to [`FileSessionHandler`](api:SilverStripe\Control\SessionHandler\FileSessionHandler) which is *non-blocking* which means that multiple concurrent requests from the same user do *not* have to wait for one another to finish.
```

Reviewer 2: Done but without the emphasis around "not"

Note that in edge case scenarios, for example if your application wants to modify a session value *based on the value that is already set* and must do so for each request, non-blocking sessions may cause unexpected results.

> If you want to use the default PHP file session handler instead, you can set the `Session.save_handler` configuration to `null`.

Reviewer: ```suggestion
If you want to use the default blocking PHP file session handler instead, you can set the `Session.save_handler` configuration to `null`.
```

```yml
SilverStripe\Control\Session:
  save_handler: null
```

You can also set it to the FQCN or injector service name of any session handler that implements [`SessionHandlerInterface`](https://www.php.net/manual/en/class.sessionhandlerinterface.php).

See the [save handler documentation](/developer_guides/cookies_and_sessions/sessions/#save-handler) for more information about the new `FileSessionHandler` including how it differs from the default PHP file session handler.

Note that if you are using [`silverstripe/hybridsessions`](https://packagist.org/packages/silverstripe/hybridsessions) or [`silverstripe/dynamodb`](https://packagist.org/packages/silverstripe/dynamodb), you can continue to use these with no change, provided you update them to the latest version.

### Other new features

## API changes

### Deprecated API

The following API has been deprecated in this release, and will be removed or replaced in a future major release.

> It's best practice to avoid usnig deprecated code where possible, but sometimes it's unavoidable. This API will all continue to be available at least until the next major release.

Reviewer 2: Blurb matches https://github.com/silverstripe/developer-docs/pull/734
